### DISCUSSION TITLE
Build time too  long when we re-run the App

### DISCUSSION THREAD

**Discussion by asnagni** [axmolengine/axmol#1885] [General]
When we re-run our App without changing (no code change), everything will build again. That a long wasted. Maybe there is an option in Axmol to avoid this situation? No sure why this happening!

- axmol version: 2.1.2
- devices test on: iPhone 11, 12 , 13, 14
- developing environments
   - NDK version: N/A
   - Xcode version: N/A
   - Visual Studio: 
      - VS version: N/A
      - MSVC version: N/A
      - Windows SDK version: N/A
   - cmake version: 
Steps to Reproduce:

1. Create an Axmol project
2. Add your files
3. Build
4. Run your App.
5. Stop the App (via Xcode) 
6. Run the App from you device (Do not use Xcode)
7. Stop the App
8. Run your App again with Xcode
9. Everything will be rebuilt again

**Comment by asnagni** [comment]
Hi guys,
   I have a way to solve this issue. Now the build is  8x to 10x faster. Only what changed will be rebuild and no rebuild of the Axmol libraries. It is as fast as cocos2d-x and in some cases faster. 
    I do believe that this should should be included in the Axmol template build. That will help convert more people to the Axmol platform .
    
 How can I share the solution that I have?
 
 Thank you.

**Comment by asnagni** [comment]
Should I create one feature request?
Use this entry?

Feature request: Suggest an idea for this project

**Comment by danialias** [comment]
Have you tested this issue with other not-Axmol projects? It may be a Xcode problem.

I found a similar question in Reddit: https://www.reddit.com/r/iOSProgramming/s/63mUKEZ630

**Comment by asnagni** [comment]
This is the document that I have to significantly improve Axmol build time. That may help anyone that would like to improve their productivity.

[AxmolBuildTime.pdf](https://github.com/axmolengine/axmol/files/14939416/AxmolBuildTime.pdf)

**Comment by rh101** [comment]
I haven't managed to reproduce this specific issue at all, with either an existing project or a new project. The application is never rebuilt after step 8, it just simply runs on the device.

Using Xcode 15.0.1

**Comment by paulocoutinhox** [comment]
Hi,

I have the same problem on Xcode and not on Android.

I see this more when add cocoapods dependency that use ObjC bridge, like RevenueCat framework.

**Comment by rh101** [comment]
This issue has started happening now, and I'm not sure what triggered it.  On rare occasions it doesn't rebuild everything, but the majority of the time it does, which means 5+ minutes of building everything when only single line is changed in the game project C++ source (not even a header change).  This shouldn't be happening at all, and it's only an issue with Xcode (VS and Android Studio builds do not have this problem).

**Reply by paulocoutinhox** [reply]
Hi, i have the same problem. What im doing is build/develop on Android Studio that dont have this problem and build for iOS/tvOS only to publish or final test.

**Reply by rh101** [reply]
> Hi, i have the same problem. What im doing is build/develop on Android Studio that dont have this problem and build for iOS/tvOS only to publish or final test.

I understand, but that's not really the issue here.  For me personally, 99%+ of my development is in Visual Studio, which allows me to be more productive than when using any other IDE.  

The issue is having to work on very specific iOS/MacOS related code, and unfortunately, there is no way to avoid Xcode for this.

I'm yet to figure out what is causing the unnecessary re-builds, even when there are *no* changes at all to the project.

**Comment by smilediver** [comment]
Question for those experiencing this problem: are you using CocoaPods in the project and are you using this hack `set_xcode_property(${APP_NAME} CONFIGURATION_BUILD_DIR "\$(BUILD_DIR)/\$(CONFIGURATION)\$(EFFECTIVE_PLATFORM_NAME)")`?

**Reply by rh101** [reply]
Yes, with this:
```
set_target_properties(${APP_NAME} PROPERTIES
        XCODE_ATTRIBUTE_CONFIGURATION_BUILD_DIR "\${BUILD_DIR}/\$(CONFIGURATION)\$(EFFECTIVE_PLATFORM_NAME)"
      )
```

Why do you ask?

**Reply by smilediver** [reply]
I believe I've never experienced rebuilding while working with Xcode project targeting macOS, but yesterday I was working with one targeting iOS, and I think I've experienced this. It's a bit hard to tell, since I was porting our unit tests to work on iOS and was changing the project left and right, so it warranted the rebuilds, but I think I also saw rebuilds when I just did some cpp changes. I didn't look in depth into that yet, so I'm not 100% sure. But I also know that the above CMake hack is really bad (see this comment https://gitlab.kitware.com/cmake/cmake/-/issues/17897#note_1337299), so I just had a suspicion that this CMake hack might be the reason.

**Reply by rh101** [reply]
You could be right, because I only started using this version of the work-around 5 months ago because of a suggestion posted in this [thread](https://github.com/axmolengine/axmol/discussions/1721), and then didn't do much work on the iOS related code for the project until a few weeks ago.  Prior to this I definitely was not experiencing the re-build issue, but also, I was using this instead:

```
set(CMAKE_XCODE_ATTRIBUTE_FRAMEWORK_SEARCH_PATHS "\${BUILD_DIR}/\$(CONFIGURATION)\$(EFFECTIVE_PLATFORM_NAME)")
```

I found this comment in my `CMakeLists.txt` too when the change was made to set `XCODE_ATTRIBUTE_CONFIGURATION_BUILD_DIR `:
```
# set XCODE_ATTRIBUTE_CONFIGURATION_BUILD_DIR instead of the CMAKE_XCODE_ATTRIBUTE_FRAMEWORK_SEARCH_PATHS,
# as it seems to be the better way do handle this
```

What I'll try over the next day or so is to go back to the `CMAKE_XCODE_ATTRIBUTE_FRAMEWORK_SEARCH_PATHS` method, and see if that fixes the rebuild issue for me.

**Reply by smilediver** [reply]
`FRAMEWORK_SEARCH_PATHS` solution doesn't work (at least not in all cases). When the target is built, then CocoaPods runs `Target-frameworks.sh` script to copy the built frameworks to the built target. In that script to find the built frameworks it uses two variables, either `BUILT_PRODUCTS_DIR` or `PODS_CONFIGURATION_BUILD_DIR` (I think one is for bundles, another is for normal frameworks, or something like that). If your used Pods happen to use `PODS_CONFIGURATION_BUILD_DIR`, then all is good, but if it uses `BUILT_PRODUCTS_DIR`, then it fails, because frameworks are built into some random Xcode's derived data dir, but `BUILT_PRODUCTS_DIR` is derived from app's `CONFIGURATION_BUILD_DIR`, which is set by CMake and points to CMake's binary dir.

It's very hard to explain this problem, cause it's complicated and with a lot of moving parts, but it's something like this:

### How Xcode normally works:

* `BUILD_DIR` = `SYMROOT` = `~/Library/Developer/Xcode/DerivedData/<random stuff>`
* `CONFIGURATION_BUILD_DIR` = `$(BUILD_DIR)/$(CONFIGURATION)$(EFFECTIVE_PLATFORM_NAME)`
* `BUILT_PRODUCTS_DIR` = `CONFIGURATION_BUILD_DIR`

Your built app and frameworks end up in `BUILT_PRODUCTS_DIR` in derived data folder, and all is good.

### Now you use CMake:

* `BUILD_DIR` = `SYMROOT` = `~/Library/Developer/Xcode/DerivedData/<random stuff>`
* `CONFIGURATION_BUILD_DIR` = `$(BUILD_DIR)/$(CONFIGURATION)$(EFFECTIVE_PLATFORM_NAME)`
* `BUILT_PRODUCTS_DIR` = `CONFIGURATION_BUILD_DIR`
* `CONFIGURATION_BUILD_DIR` for your app = `some/cmake/path`
* `BUILT_PRODUCTS_DIR` for your app = `CONFIGURATION_BUILD_DIR` for your app, so `some/cmake/path`

Your built app is in `some/cmake/path`, frameworks are built with normal `BUILT_PRODUCTS_DIR` so ends up in derived data, install script is run in app target so it uses app's `BUILT_PRODUCTS_DIR`, so tries to copy framework at `BUILT_PRODUCTS_DIR/framework`, ie at `some/cmake/path/framework` and fails.

I've wasted like two days to find a different way than `CONFIGURATION_BUILD_DIR`, but nothing works. Everything points that the end solution is to force frameworks to build into `some/cmake/path`, and it's pretty easy to force that, but then there's a big complication of setting up the app's target to find the frameworks, and requires modifying something after `pod install`: either modifying app's Xcode project, or modifying CocoaPods xcconfig files, which sounds like a terrible hack.

A handy thing: https://pewpewthespells.com/blog/buildsettings.html

**Reply by rh101** [reply]
Thanks for the extra information.  After doing some more testing with different combinations of settings, nothing would work to resolve the re-build issue, which even happens when no code is modified.

There doesn't seem to be away to avoid setting `CONFIGURATION_BUILD_DIR`; if it's not set, then the build fails attempting to locate something related to the frameworks imported via CocoaPods.  

It doesn't seem to be an issue with Axmol, but more with Xcode/CMake, one that I hope they address in the near future.

**Reply by smilediver** [reply]
Did you find that `CONFIGURATION_BUILD_DIR` is causing the rebuild issue, or you couldn't build the project without it?

If `CONFIGURATION_BUILD_DIR` is the issue, then I think the problem is in CocoaPods using `BUILT_PRODUCTS_DIR` for copying the frameworks and forcing us to use `CONFIGURATION_BUILD_DIR` hack. If this is indeed the problem, we could make a small repro project and raise the issue with CocoaPods devs.

**Reply by rh101** [reply]
My project doesn't build without setting the `CONFIGURATION_BUILD_DIR`, but hen I have some time I'll take out the references to pod frameworks and try it again.

For the time being, we should be able to try this out with a simple test project.

**Reply by smilediver** [reply]
So I've tried to reproduce this rebuild problem with a minimal project. I just created a new macOS app project in Xcode, deleted Xcode project itself and set it up to generate through CMake. With CMake generated project everything works fine, I can build the app, and changing one file doesn't rebuild everything. But, if I create Xcode workspace including that project and build through workspace, then changing one file always rebuilds whole project. It rebuilds with `CONFIGURATION_BUILD_DIR` hack and without it, and it makes sense, since CMake always sets it anyway. My guess is that building through project uses the project's `SYMROOT` value, which is set by CMake, and building through workspace Xcode sets up a different value for `SYMROOT` (perhaps which is supposed to be shared between projects in the workspace).

And I also have found a workaround for the rebuild problem. Go to workspace settings, click advanced, and choose legacy build location. After this, rebuilding problem was gone. I haven't tried it with a real project yet, but it seems promising. Though, this settings seems to be set per user.

**Reply by paulocoutinhox** [reply]
I will test here in my game: https://paulocoutinho.pages.dev/gamemania/home/

**Reply by paulocoutinhox** [reply]
Very nice, it works for me.

<img width="1512" alt="image" src="https://github.com/user-attachments/assets/f0047e36-4a9f-44cc-bc43-1bc833e227b0">

**Reply by paulocoutinhox** [reply]
There is any way to set it from podfile? Since the workspace is generate by it?

**Reply by paulocoutinhox** [reply]
I know that there is a command line for this:

```
xcodebuild -workspace Foo.xcworkspace -scheme Bar -configuration Release -archivePath /path/to/Foo.xcarchive clean archive -UseModernBuildSystem=NO
```

**Reply by paulocoutinhox** [reply]
when you update the Xcode->File->Project Settings->Build System = Legacy Build System, it updates Projectname.xcodeproj/project.xcworkspace/xcshareddata/WorkspaceSettings.xcsettings.

```
<dict>
	<key>BuildSystemType</key>
	<string>Original</string>
</dict>
```

when you set it to "New Build System", this file does not have this block

**Reply by paulocoutinhox** [reply]
Based on this, i add this to my makefile rule:

```
build-ios:
	rm -rf build_ios_arm64/
	axmol build -p ios -a arm64 -c
	cd build_ios_arm64 && pod install
	cd build_ios_arm64/${PROJ}.xcworkspace && mkdir -p xcuserdata/$(USER).xcuserdatad && \
	plutil -create xml1 xcuserdata/$(USER).xcuserdatad/WorkspaceSettings.xcsettings && \
	plutil -insert BuildSystemType -string Original xcuserdata/$(USER).xcuserdatad/WorkspaceSettings.xcsettings && \
	plutil -insert BuildLocationStyle -string UseTargetSettings xcuserdata/$(USER).xcuserdatad/WorkspaceSettings.xcsettings
```

**Reply by smilediver** [reply]
Build location and build system are two different things. As far as I can tell, this rebuilding issue is caused by the build location and not the build system choice. Also the legacy build system is deprecated and will be removed.

**Comment by asnagni** [comment]
Hi,
  1) Yes using cocoapods.
  2) Yes using the hack :     
    set_target_properties(${APP_NAME} PROPERTIES XCODE_ATTRIBUTE_CONFIGURATION_BUILD_DIR "\${BUILD_DIR}/\$(CONFIGURATION)\$(EFFECTIVE_PLATFORM_NAME)”)


Thank you,
Stay safe

**Reply by paulocoutinhox** [reply]
I have this too:

```
if(APPLE)
    if(TVOS)
        set_target_properties(${APP_NAME} PROPERTIES XCODE_ATTRIBUTE_CONFIGURATION_BUILD_DIR "\${BUILD_DIR}/\$(CONFIGURATION)\$(EFFECTIVE_PLATFORM_NAME)")
        configure_file("${CMAKE_CURRENT_SOURCE_DIR}/Podfile.tvos" "${CMAKE_CURRENT_BINARY_DIR}/Podfile" COPYONLY)
    elseif(IOS)
        set_target_properties(${APP_NAME} PROPERTIES XCODE_ATTRIBUTE_CONFIGURATION_BUILD_DIR "\${BUILD_DIR}/\$(CONFIGURATION)\$(EFFECTIVE_PLATFORM_NAME)")
        configure_file("${CMAKE_CURRENT_SOURCE_DIR}/Podfile.ios" "${CMAKE_CURRENT_BINARY_DIR}/Podfile" COPYONLY)
    endif()
endif()
```

But i have the same problems. When i change one line of code, it rebuild near everything and take a long time too build.

**Comment by asnagni** [comment]
Hi there,
   	Yes I remember this issue when I migrated to Axmol. It can cost you lot of valuable time. I posted a solution to solve this problem and it works 100%. Since then I don’t have to worry anymore about it anymore. I posted a pdf document that show you how to mitigate the problem. Take a look at it, it will save you lot of time.
	When The Axmol will come up with a better solution, you can swap to their solution/proposal.

Thank you,
Stay safe

**Comment by asnagni** [comment]
Hi ,
	I couldn’t build without the hack.

Stay safe.

**Comment by rh101** [comment]
Just an observation, when building a release version with debug info, then changing a single line of source code, the re-build issue never happened.  It simply re-compiled the single file that was affected by the change, and then the final package.  This was attempted multiple times, even with header file changes, where only the affected files referencing that header were re-compiled, and at no point did it rebuild the entire engine code etc..  

The work-around of using the legacy build location wasn't used.  This project is the same workspace that is having issues with the debug build, causing re-builds any time the project build or run command is used, even when there are no changes to the code.

**Reply by smilediver** [reply]
Wow, I can confirm that on my repro project. Switching to `Release`, `MinSizeRel` and `RelWithDebInfo` doesn't rebuild everything.

**Reply by smilediver** [reply]
So the reason why rebuild is triggered is because the response file containing common compiler options is rewritten. What triggers its rewrite only in debug builds I have no idea. An example bellow where I modify `other.m` file:
![image](https://github.com/user-attachments/assets/c9a14654-8034-4e47-898a-2c162906d7f0)

And another workaround seems to be to turn off "Use Response Files" build setting. Haven't tested on real project yet, but something like this should work:
```cmake
set_target_properties(${APP_NAME} PROPERTIES XCODE_ATTRIBUTE_CLANG_USE_RESPONSE_FILE No)
```

**Reply by rh101** [reply]
I just tried build with this:

`set_target_properties(${APP_NAME} PROPERTIES XCODE_ATTRIBUTE_CLANG_USE_RESPONSE_FILE No)`

The rebuild still occurs, but not everything gets rebuilt, which is good, and this is all without any changes made to the project.

**Comment by asnagni** [comment]
Hi,
 	I did some tests and don’t see the issue if I don’t use cocoapods. Are you experiencing the same thing?

Thanks

**Reply by smilediver** [reply]
> Hi, I did some tests and don’t see the issue if I don’t use cocoapods.

This has nothing to do with CocoaPods, I can reproduce this with a simple project without CocoaPods, see https://github.com/axmolengine/axmol/discussions/1885#discussioncomment-10098170. It seems that rebuilding manifests only if you're building through Xcode workspace. If you build through Xcode project then everything is fine.

**Comment by asnagni** [comment]
Ok thank you for the clarifications.

Stay safe

**Comment by AlexandreK38** [comment]
Hi! Jumping in this thread as I face the same issue, has anyone found a fix yet?

**Reply by rh101** [reply]
Could you possibly try the suggestion by @smilediver from this [post](https://github.com/axmolengine/axmol/discussions/1885#discussioncomment-10115862)?

I'll be able to test this out later this week, but I do recall that setting the build to release with debug info was a work-around for me.  The other option is as @asnagni suggested, using ccache.

**Reply by AlexandreK38** [reply]
smilediver's solution seems to improve, what I don't understand is why xcode compiles each file again though, including the axmol libs ones

**Reply by rh101** [reply]
@smilediver pointed out the possible cause of the rebuilds in a previous post up above, but I personally have no idea why it happens.  It really has been a huge waste of time, especially given the size of the project.  The only way I have avoided it is by switching to RelWithDebInfo for the build, but in order to debug the code reliably, all optimizations would need to be turned off for the release build.

**Reply by AlexandreK38** [reply]
it seems to compile only what we need now with `Use Responsive Files` to `No` in the project (not target) and in the ax_libs sub project too. Targets inherit this value

**Reply by rh101** [reply]
What do you mean by "in the project (not target) and in the ax_libs sub project too"? Is this via the line:

`set_target_properties(${APP_NAME} PROPERTIES XCODE_ATTRIBUTE_CLANG_USE_RESPONSE_FILE No)`

Or is it something you set via the Xcode UI, or a different line in cmake?

**Reply by AlexandreK38** [reply]
Yeah I set it using the Xcode UI but it should be the same as 
`set_target_properties(${APP_NAME} PROPERTIES XCODE_ATTRIBUTE_CLANG_USE_RESPONSE_FILE No)`

As side note I didn't want to build the xcode project each time as the Cmakefile does, because with Cocoapods, Firebase, and few other tools have to be setup each time so I generate one Axmol project with subproject, extract the `ax_libs.xcodeproj`, tweak the links in it to be relative ones and use it in my main game xcodeproj, which is never regenerated. I always compile using Xcode

**Reply by rh101** [reply]
> As side note I didn't want to build the xcode project each time as the Cmakefile does, because with Cocoapods, Firebase, and few other tools have to be setup each time

What do you mean by "setup each time"?  The reason I ask is because how I have it set up is that `axmol build ****` is used to re-generate the xcode project, then I run `pod install`, and that's it.  Everything is setup automatically, and this includes the Firebase cpp framework which isn't part of the Cocoapod setup.  I've never had to setup anything manually in the Xcode IDE at all, if that's what you meant by "setup each time".

**Reply by AlexandreK38** [reply]
Even with Cocos2d before we used a subproject and do not build using the axmol commands; when we switched to Axmol I had a look to build like you do, with podfile and firebase it's ok, but I have some third party ruby scripts to run that modify the project and look at the pods too, it was a bit messy to make it work and run it each time after axmol build command. My CMakefile is  more or less "ready" for axmol command but we don't use it and only use Xcode to build. I prefer this way and the axmol sources and `ax_libs` xcode project are in a common github submodule that we use for all our games.

**Comment by asnagni** [comment]
Hi AlexandreK38,
  As far as I remember, I don’t think there was a definitive solution to this issue but I may be wrong. But to avoid this issue we are using ccache, so we don’t think about it anymore. 

@rh101 would you have more information on this?

I hope that helps.

**Comment by asnagni** [comment]
Hi AlexandreK38,
   We faced this issue when we first migrated to Axmol. After investigating we were not able to find the root cause so we implemented the ccache solution. Since, nobody is thinking about it since it is working perfectly for us now. Perhaps @rh101 may have some answers on that issue, you may want to ask him.


Thank you,
Stay safe
